chaoscenter/authentication/pkg/project/repository.go

```diff
@@ -69,51 +69,75 @@ func (r repository) GetProjects(query bson.D) ([]*entities.Project, error) {
 }
 
 // GetProjectsByUserID returns a project based on the userID
-func (r repository) GetProjectsByUserID(userID string, isOwner bool) ([]*entities.Project, error) {
+func (r repository) GetProjectsByUserID(request *entities.ListProjectRequest) (*entities.ListProjectResponse, error) {
 	var projects []*entities.Project
-	query := bson.D{}
 
-	if isOwner == true {
-		query = bson.D{
-			{"members", bson.D{
-				{"$elemMatch", bson.D{
-					{"user_id", userID},
-					{"role", bson.D{
-						{"$eq", entities.RoleOwner},
-					}},
-				}},
-			}}}
-	} else {
-		query = bson.D{
-			{"is_removed", false},
-			{"members", bson.D{
-				{"$elemMatch", bson.D{
-					{"user_id", userID},
-					{"$and", bson.A{
-						bson.D{{"invitation", bson.D{
-							{"$ne", entities.PendingInvitation},
-						}}},
-						bson.D{{"invitation", bson.D{
-							{"$ne", entities.DeclinedInvitation},
-						}}},
-						bson.D{{"invitation", bson.D{
-							{"$ne", entities.ExitedProject},
-						}}},
-					}},
-				}},
-			}}}
+	// Construct the pipeline
+	var pipeline mongo.Pipeline
+
+	pipeline = append(pipeline, utils.CreateMatchStage(request.UserID))
+
+	if request.Filter != nil {
+		filterStages := utils.CreateFilterStages(request.Filter, request.UserID)
+		pipeline = append(pipeline, filterStages...)
+	}
+
+	sortStage := utils.CreateSortStage(request.Sort)
+	if len(sortStage) > 0 {
+		pipeline = append(pipeline, sortStage)
 	}
 
-	result, err := r.Collection.Find(context.TODO(), query)
+	// count stage
+	var pipelineCount mongo.Pipeline
+	pipelineCount = append(pipeline, bson.D{
+		{"$count", "totalNumberOfProjects"},
+	})
+
+	countCursor, err := r.Collection.Aggregate(context.TODO(), pipelineCount)
 	if err != nil {
 		return nil, err
 	}
-	err = result.All(context.TODO(), &projects)
+
+	// Extract count result
+	var countResult []bson.M
+	if err := countCursor.All(context.TODO(), &countResult); err != nil {
+		return nil, err
+	}
+
+	var totalNumberOfProjects int64
+	if len(countResult) > 0 {
+		totalNumberOfProjects = int64(countResult[0]["totalNumberOfProjects"].(int32))
+	}
+
+	paginationStages := utils.CreatePaginationStage(request.Pagination)
+	for _, stage := range paginationStages {
+		pipeline = append(pipeline, stage)
+	}
+
+	cursor, err := r.Collection.Aggregate(context.TODO(), pipeline)
 	if err != nil {
 		return nil, err
 	}
+	defer cursor.Close(context.TODO())
+
+	// Iterate over the cursor and decode the results into projects
+	for cursor.Next(context.TODO()) {
+		var project entities.Project
+		if err := cursor.Decode(&project); err != nil {
```

Comment:
Instead of looping over cursor, cursor.All can be used to decode multiple elements in the map